Fix: dot-location errors with parenthesized objects (fixes #11868)

mdjermanovic opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[X] Bug fix #11868

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**

Fixed `dot-location` rule (`"object"` option) because it produces syntax errors with parenthesized objects. 

Parentheses around `MemberExpression.object` node are now treated as a part of the 'object'.

For example, the following was a false positive:

```js
( 
  object
).
property
```

and the auto-fix produced syntax error:

```js
( 
  object.
)
property
```

With this PR, the example above does not have a warning.

Also, with this PR auto-fix works correctly for similar cases that do have warnings. For example, this:

```js
( 
  object
)
.property
```

will be fixed to:

```js
( 
  object
).
property
```

I've also added several test cases unrelated to this issue.

**Is there anything you'd like reviewers to focus on?**

* If the first token before the `MemberExpression.property` is not a `dot` token, the rule will ignore that node. I believe there is no need to support this case ([post](https://github.com/eslint/eslint/issues/11868#issuecomment-506919269)).
* The rule doesn't check if the `dot` is right after the `object` (or parens now) / right before the `property`. This is valid syntax: `a   .   b` and produces no `dot-location` warnings. Perhaps there could be an additional option to 'attach' the dot, I could open an issue if this wasn't already discussed.

not-an-aardvark commented on lib/rules/dot-location.js:
```diff
@@ -54,29 +54,36 @@ module.exports = {
-            const textBeforeDot = sourceCode.getText().slice(obj.range[1], dot.range[0]);
+
+            /* istanbul ignore if */
+            if (astUtils.isNotDotToken(dot)) {
```
What's the purpose of this conditional? Is it just a sanity check?

mdjermanovic replied:
Yes, it's a bit strange now. 

This should never happen, I couldn't find an example in proposals, typescript etc. It's basically inherited from the previous version, the purpose was most likely to check for `[]` (computed key), which isn't necessary now. 

But, it was also a check for an unknown syntax, so I wasn't sure should it stay or not. 

Anyway, I completely agree that the purpose of this line is not clear to a reader, and it's better to do something about it. Should I remove the check, or add a comment?

mdjermanovic replied:
I've added the comment.

If it would be better, I can remove the whole conditional, too.

kaicataldo replied:
Unless we can find a test case where this guard is necessary, I’d be more inclined to remove it for future maintainers’ sake.

mdjermanovic commented:
I'll be away for two weeks from next Tuesday. If the `isNotDotToken` is a blocking issue I could remove that conditional now, to not wait that long. In any case, please don't close the PR if I'm not responding in next couple of weeks :)

kaicataldo commented:
My vote would be to remove it, but I also don’t want to hold up landing this. If you can remove it, that sounds good to me! Otherwise, I’d be fine with merging this and revisiting that. Thanks for working on this!

mdjermanovic commented:
It's removed now. 

It looks like this conditional (in the form: `if (dot.type === "Punctuator" && dot.value === "."`) was indeed added to check computed properties in 45cda23 (PR #2506 which also fixed an unrelated or semi-related #2504).

I didn't remove `isDotToken()` and `isNotDotToken()` functions from `ast_utils`. They were added in the first commit, both are unused now but they have their own test cases and might be useful for something else in the future.

I've also added more test cases with computed properties.
